libvast/src/data.cpp

```diff
@@ -274,92 +274,83 @@ const data* get(const data& d, const offset& o) {
   return nullptr;
 }
 
-namespace {
-
-template <class Iterator>
-vector flatten(Iterator& f, Iterator l) {
-  vector xs;
-  xs.reserve(l - f);
-  for (; f != l; ++f) {
-    auto&& x = *f;
-    if (auto v = caf::get_if<vector>(&x)) {
-      auto begin = Iterator{v->begin()};
-      auto end = Iterator{v->end()};
-      auto ys = flatten(begin, end);
-      xs.insert(xs.end(),
-                std::make_move_iterator(ys.begin()),
-                std::make_move_iterator(ys.end()));
+caf::optional<vector> flatten(const vector& xs, const record_type& t) {
+  if (xs.size() != t.fields.size())
+    return {};
+  vector result;
+  for (auto i = 0u; i < xs.size(); ++i) {
```

Comment:
`0u` is an `unsigned`, so this constantly needs to promote a 32-bit value to 64-bits when comparing to `xs.size()` or passing to `xs[i]`. Since there's no reason for `i` to have any type other than `size_t` I'd spell it out here: `for (size_t i = 0; ...)`.